src/contracts/ProofOfPhysicalAddress.contract.js

```diff
@@ -0,0 +1,107 @@
+
+    const branch = helpers.getBranch(netId)
+    if (branch !== 'core') {
+      throw new Error(`ProofOfPhysicalAddress contract not deployed on network "${branch}"`)
+    }
+    let proofOfPhysicalAddressAbi = await helpers.getABI(branch, 'ProofOfPhysicalAddress')
+    this.instance = new web3_10.eth.Contract(proofOfPhysicalAddressAbi, PROOF_OF_PHYSICAL_ADDRESS)
+
+    this.getPhysicalAddressesOfWalletAddressArray = this.getPhysicalAddressesOfWalletAddressArray.bind(this)
+    this.getAllEvents = this.getAllEvents.bind(this)
+    this.getPhysicalAddressesByWalletAddressAndKeccakIdentifierArray = this.getPhysicalAddressesByWalletAddressAndKeccakIdentifierArray.bind(
+      this
+    )
+  }
+
+  /**
+   * Given a wallet address array, return a promise that resolves to an array of physical addresses from
+   * PoPA contract.
+   * @param {String[]} walletAddressArray
+   * @return {Promise}
+   */
+  async getPhysicalAddressesOfWalletAddressArray(walletAddressArray) {
+    let result = []
+    try {
+      const registeredAddressEvents = await this.getAllEvents(REGISTER_ADDRESS_EVENT_NAME)
+      const allPhysicalAddressesPromises = walletAddressArray.map(walletAddress => {
+        let physicalAddresses = null
+        let keccakIdentifiers = registeredAddressEvents
+          .filter(event => event.returnValues.wallet === walletAddress)
+          .map(event => event.returnValues.keccakIdentifier)
+
+        if (keccakIdentifiers.length > 0) {
+          physicalAddresses = this.getPhysicalAddressesByWalletAddressAndKeccakIdentifierArray(
+            walletAddress,
+            keccakIdentifiers
+          ).then(physicalAddresses => {
+            return physicalAddresses.length > 0 ? physicalAddresses : null
+          })
+        }
+        return Promise.resolve(physicalAddresses)
+      })
+      result = await Promise.all(allPhysicalAddressesPromises)
+    } catch (e) {
+      console.error(`Error in getPhysicalAddressesOfWalletAddressArray`, e)
+    }
+    return result
+  }
+
+  /**
+   * Given a walletAddress and an array of keccakIdentifiers, return a promise that resolves to an array
+   * of the corresponding physical addresses, or an empty array.
+   * @param  {String}  walletAddress
+   * @param  {String[]}  keccakIdentifierArray
+   * @return {Promise}
+   */
+  async getPhysicalAddressesByWalletAddressAndKeccakIdentifierArray(walletAddress, keccakIdentifierArray) {
+    let promises = keccakIdentifierArray.map(keccakIdentifier => {
+      return this.instance.methods
+        .userAddressByKeccakIdentifier(walletAddress, keccakIdentifier)
+        .call()
+        .then(async addressStatusTuple => {
+          const addressFound = addressStatusTuple[0]
+          const addressIndex = addressStatusTuple[1]
+          const addressConfirmed = addressStatusTuple[2]
+          // If addressIndex === 0, it also means the address was not found
+          if (addressFound === false || addressIndex === 0) {
```

Comment:
@unjapones I think this condition is wrong because zero index is not the same as `address not found`. Here we can see that the found address can have zero index: https://github.com/poanetwork/poa-popa/blob/cdbe618cf00e2c45b3c67d88a0d84a065a2fe480/blockchain/contracts/ProofOfPhysicalAddress.sol#L183 - the variable `ai` may be `0` even when `found` flag is `true`.

Reply:
You are right: the 2nd part of the condition check is redundant and the 1st part `addressFound === false` is enough.
I'll remove the redundant verification and the comment in the previous line (I believe the code is clear enough).